src/main/java/com/algolia/search/ConfigBase.java

```diff
@@ -1,44 +1,136 @@
 package com.algolia.search;
 
+import com.algolia.search.util.AlgoliaUtils;
 import java.util.HashMap;
 import java.util.List;
+import java.util.Objects;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.ForkJoinPool;
+import javax.annotation.Nonnull;
 
-/** * Algolia's clients common configuration */
+/** Algolia's clients common configuration */
 @SuppressWarnings("WeakerAccess")
-public abstract class ConfigBase {
-
-  /**
-   * Creates an default configuration with the forkJoinPool.
-   *
-   * @param applicationID The Application ID
-   * @param apiKey The API Key.
-   */
-  ConfigBase(String applicationID, String apiKey) {
-    this(applicationID, apiKey, ForkJoinPool.commonPool());
-  }
+abstract class ConfigBase {
+
+  private static final String javaVersion = System.getProperty("java.version");
+  private static final String clientVersion = "3.0.0.0";
+  private final String applicationID;
+  private final String apiKey;
+  private final HashMap<String, String> defaultHeaders;
+  private final int batchSize;
+  private final Integer readTimeOut;
+  private final Integer writeTimeOut;
+  private final Integer connectTimeOut;
+  private final List<StatefulHost> defaultHosts;
+  private final List<StatefulHost> customHosts;
+  private final ExecutorService executor;
 
-  /**
-   * Creates a advanced configuration with a custom ExecutorService.
-   *
-   * @param applicationID The Application ID
-   * @param apiKey The API Key
-   * @param executor The Executor service
-   */
-  ConfigBase(String applicationID, String apiKey, ExecutorService executor) {
-    this.applicationID = applicationID;
-    this.apiKey = apiKey;
-    this.executor = executor;
-    this.batchSize = 1000;
-    this.defaultHeaders = new HashMap<>();
-    this.defaultHeaders.put(Defaults.ALGOLIA_APPLICATION_HEADER, getApplicationID());
-    this.defaultHeaders.put(Defaults.ALGOLIA_KEY_HEADER, getApiKey());
-    this.defaultHeaders.put(
-        Defaults.USER_AGENT_HEADER,
-        String.format("Algolia for Java (%s); JVM (%s)", clientVersion, javaVersion));
-    this.defaultHeaders.put(Defaults.ACCEPT_HEADER, Defaults.APPLICATION_JSON);
-    this.defaultHeaders.put(Defaults.ACCEPT_ENCODING_HEADER, Defaults.CONTENT_ENCODING_GZIP);
+  /** Config base builder to ensure the immutability of the configuration. */
+  public abstract static class Builder<T extends Builder<T>> {
+
+    private final String applicationID;
+    private final String apiKey;
+    private final HashMap<String, String> defaultHeaders;
+    private int batchSize;
+    private Integer readTimeOut;
+    private Integer writeTimeOut;
+    private Integer connectTimeOut;
+    private List<StatefulHost> defaultHosts;
+    private List<StatefulHost> customHosts;
```

Comment:
I suggest to have just: `private List<StatefulHost> hosts;`. No "default" vs "custom" ones.